src/run/headers.ts

```diff
@@ -281,13 +283,24 @@ export const setCacheControlHeaders = (
   }
 }
 
-export const setCacheTagsHeaders = (headers: Headers, requestContext: RequestContext) => {
-  if (
-    requestContext.responseCacheTags &&
-    (headers.has('cache-control') || headers.has('netlify-cdn-cache-control'))
-  ) {
+export const setCacheTagsHeaders = (
+  headers: Headers,
+  request: Request,
+  requestContext: RequestContext,
+) => {
+  if (!headers.has('cache-control') && !headers.has('netlify-cdn-cache-control')) {
+    return
+  }
+
+  if (requestContext.responseCacheTags) {
     headers.set('netlify-cache-tag', requestContext.responseCacheTags.join(','))
+    return
   }
+
+  const key = new URL(request.url).pathname
+  const cacheTag = `_N_T_${key === '/index' ? '/' : encodeURI(key)}`
+  console.log('setCacheTagsHeaders', 'netlify-cache-tag', key)
+  headers.set('netlify-cache-tag', cacheTag)
```

Comment:
deriving cache tag from request pathname is not reliable

- there could be rewrites ( https://nextjs.org/docs/app/api-reference/config/next-config-js/rewrites )
- json responses for Pages Router would be something like `_next/data/build_id/<route>.json`, so would not share cache tags with html routes
- there could be trailingSlash option enabled ( https://nextjs.org/docs/app/api-reference/config/next-config-js/trailingSlash )

(just to name a few that immediately comes to mind - there likely is much more)

Instead we should figure out how to capture those more reliably in cache handler, so logic here is not really changed (but we can keep simplified breaking up of condition with early bail so it's more readable overall)